tempesta_fw/http.c

```diff
@@ -446,25 +488,27 @@ tfw_http_req_cache_cb(TfwHttpReq *req, TfwHttpResp *resp, void *data)
 		 */
 		tfw_connection_send(req->conn, (TfwMsg *)resp);
 	} else {
-		/* Dispatch request to an appropriate server. */
-		TfwConnection *conn = tfw_sched_get_srv_conn((TfwMsg *)req);
-		if (!conn) {
-			TFW_ERR("Unable to find a backend server\n");
-			goto send_404;
-		}
 		r = tfw_http_sticky_req_process((TfwHttpMsg *)req);
 		if (r < 0) {
 			goto send_500;
 		} else if (r > 0) {
 			/* Response sent, nothing to do */
-			tfw_http_msg_free((TfwHttpMsg *)req);
+			tfw_http_conn_msg_free((TfwHttpMsg *)req);
 			return;
 		}
 		if (tfw_http_adjust_req(req))
 			goto send_500;
 
-		/* Add request to the connection. */
+		/* Dispatch request to an appropriate server. */
+		conn = tfw_sched_get_srv_conn((TfwMsg *)req);
+		if (conn == NULL) {
+			TFW_ERR("Unable to find a backend server\n");
+			goto send_404;
+		}
```

Comment:
Schedulers should make a decision on plain unmodified request, so the scheduling should be done before any request mangling.